### PR TITLE
optimize the metrics test

### PR DESCRIPTION
**Pull request by zouyee**
Signed-off-by: zouyee <zounyee1989@gmail.com>

### 1. Why is this pull request needed and what does it do?
To optimize the monitoring test, we should not simply confirm the existence of metrics, but compare the results of metrics deeply.

### 2. Which issues (if any) are related?
https://github.com/coredns/coredns/issues/3579

### 3. Which documentation changes (if any) need to be made?
NONE

### 4. Does this introduce a backward incompatible change or deprecation?
NONE

### CODE REVIEW

**Comment by miekg** [comment]
fails travis

**Comment by zouyee** [comment]
@miekg PTAL

File: plugin/etcd/etcd.go [hunk 1]
```diff
@@ -178,7 +178,7 @@ func (e *Etcd) TTL(kv *mvccpb.KeyValue, serv *msg.Service) uint32 {
 
 // shouldInclude returns true if the service should be included in a list of records, given the qType. For all the
 // currently supported lookup types, the only one to allow for an empty Host field in the service are TXT records
-// which resolve directly.  If a TXT record is being resolved by CNAME, then we expect the Host field to have a
+// which resolve directly. If a TXT record is being resolved by CNAME, then we expect the Host field to have a
```
Comment (miekg) [inline plugin/etcd/etcd.go]: unrelated change, please revert

File: plugin/metrics/metrics.go [hunk 2]
```diff
@@ -96,7 +95,7 @@ func (m *Metrics) ZoneNames() []string {
 
 // OnStartup sets up the metrics on startup.
 func (m *Metrics) OnStartup() error {
-	ln, err := reuseport.Listen("tcp", m.Addr)
+	ln, err := net.Listen("tcp", m.Addr)
```
Comment (miekg) [inline plugin/metrics/metrics.go]: unrelated change, please revert

File: plugin/test/scrape.go [hunk 3]
```diff
@@ -62,8 +65,108 @@ type (
 		Count   string            `json:"count"`
 		Sum     string            `json:"sum"`
 	}
+	// DnsMetrics is metrics for dns
```
Comment (miekg) [inline plugin/test/scrape.go]:
```suggestion
	// DnsMetrics contain the metrics we scrape from CoreDNS.
```

File: plugin/test/scrape.go [hunk 4]
```diff
@@ -62,8 +65,108 @@ type (
...
+	DnsMetrics map[string]model.Samples
 )
 
+// NewDnsMetrics returns new metrics which are initialized.
```
Comment (miekg) [inline plugin/test/scrape.go]: this doesn't look like a very useful function...

File: plugin/test/scrape.go [hunk 5]
```diff
@@ -62,8 +65,108 @@ type (
...
+func NewDnsMetrics() DnsMetrics {
+	result := make(DnsMetrics)
+	return result
+}
+
+// GrabDnsMetrics retrieve metrics from the dns using a simple GET over htt
+// Currently only used in integration tests.
+func GrabDnsMetrics(url string) (DnsMetrics, error) {
```
Comment (miekg) [inline plugin/test/scrape.go]:
```suggestion
func scrape(url string) (DnsMetrics, error) {
```

File: plugin/test/scrape.go [hunk 6]
```diff
@@ -62,8 +65,108 @@ type (
...
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return DnsMetrics{}, err
+	}
+	return parseDnsMetrics(string(body))
+}
+
+func parseDnsMetrics(data string) (DnsMetrics, error) {
```
Comment (miekg) [inline plugin/test/scrape.go]: this intermediate function can be removed.

File: plugin/test/scrape.go [hunk 7]
```diff
@@ -62,8 +65,108 @@ type (
...
+	result := NewDnsMetrics()
+	if err := parseMetrics(data, &result); err != nil {
+		return DnsMetrics{}, err
+	}
+	return result, nil
+}
+
+// parseMetrics parses Metrics from data returned from prometheus endpoint
+func parseMetrics(data string, output *DnsMetrics) error {
```
Comment (miekg) [inline plugin/test/scrape.go:92]: parseMetrics should return a DnsMetrics

File: plugin/test/scrape.go [hunk 8]
```diff
@@ -62,8 +65,108 @@ type (
...
+			continue
+		}
+		for _, metric := range v {
+			name := string(metric.Metric[model.MetricNameLabel])
+			(*output)[name] = append((*output)[name], metric)
+		}
+	}
+}
+
+// GetDnsMetrics returns config related metrics from the dns, filtered to t
```
Comment (miekg) [inline plugin/test/scrape.go]: what's `t`?

File: plugin/test/scrape.go [hunk 9]
```diff
@@ -62,8 +65,108 @@ type (
...
+func GetDnsMetrics(filterMetricNames map[string]interface{}, url string) (DnsMetrics, error) {
+	// grab Kubelet metrics
```
Comment (miekg) [inline plugin/test/scrape.go]: comment doesn't match what's done here

File: plugin/test/scrape.go [hunk 10]
```diff
@@ -62,8 +65,108 @@ type (
...
+	for _, samples := range filtered {
+		for _, sample := range samples {
+			sample.Timestamp = 0
+		}
+	}
+	return filtered, nil
+}
+
+// NewCacheSample return the pointer of model.Sample
+func NewCacheSample(name model.LabelValue, source model.LabelSet) *model.Sample {
```
Comment (miekg) [inline plugin/test/scrape.go]: I don't understand the naming here, what's a CacheSample?

File: test/metrics_test.go [hunk 11]
```diff
@@ -29,112 +30,114 @@ example.com:0 {
...
-		t.Fatal(err)
+        if err != nil {
+		t.Fatalf("Could not get CoreDNS serving instance: %s", err)
 	}
-	// TODO(miek): make the auto sleep even less.
-	time.Sleep(1100 * time.Millisecond) // wait for it to be picked up
 
 	m := new(dns.Msg)
-	m.SetQuestion("www.example.org.", dns.TypeA)
+	m.SetQuestion("google.com.", dns.TypeA)
```
Comment (miekg) [inline test/metrics_test.go]: use (or don't change www.example.org)

File: test/metrics_test.go [hunk 12]
```diff
@@ -29,112 +30,114 @@ example.com:0 {
...
 	}
-
-	data = test.Scrape("http://" + metrics.ListenAddr + "/metrics")
-	got, _ = test.MetricValueLabel(metricName, "example.org.", data)
-
-	if got != "1" {
-		t.Errorf("Expected value %s for %s, but got %s", "1", metricName, got)
+	t.Logf("current addr： %s\n", metrics.ListenAddr)
+	// compare to expected
+	if diff := cmp.Diff(expect, actual); diff != "" {
```
Comment (miekg) [inline test/metrics_test.go]: cmd.Diff can be handy, but the old test clearly spelled out what was wrong, now this becomes more complex. I like the old way better.

File: test/metrics_test.go [hunk 13]
```diff
@@ -29,112 +30,114 @@ example.com:0 {
...
-		prometheus localhost:0
-	}
+	corefile := `example.org:0 {
+	forward . 8.8.8.8:53
+	prometheus localhost:0
+}
 `
-
-	i, err := CoreDNSServer(corefile)
+	srv, err := CoreDNSServer(corefile)
```
Comment (miekg) [inline test/metrics_test.go]: this renaming balloons this PR, please don't do that.

**Review by miekg — CHANGES_REQUESTED** [review]
this needs to be heavily minimized and only contain the changes that are actually needed

File: plugin/test/scrape.go (hunk 7 above)
Comment (miekg) [reply plugin/test/scrape.go:92]: this is still not done

**Pull request closed without merging**